[libunwind] Fix execution flow imbalance when using C++ Exceptions

medismailben opened the pull request:
This patch should fix tracing support when using C++ Exceptions.

When unwinding C++ Exceptions, libunwind would have a first searching phase to detect its caller landing pad using the personality function, and a second phase that returns to the landing pad and restore the register context by also updating the link register to point to the landing pad address found inthe first phase.

Since it changes the link register value and returns all the frames that have been called in libunwind, it causes an imbalance in the execution flow which breaks [Apple Processor Trace](https://developer.apple.com/documentation/xcode/analyzing-cpu-usage-with-processor-trace) analysis and affects its clients, like Instruments.app.

This patch addresses the issue by generating the right amount of `ret` instructions for every function called in libcxx & libunwind to rebalance the execution flow, right before returning to the catch block.

rdar://163739950

ojhunt commented on libunwind/src/assembly.h:
```diff
@@ -132,6 +133,10 @@
 
 #if defined(__APPLE__)
 
+#if (defined(__aarch64__) || defined(__arm64__) || defined(__arm64e__))
```
nit: don't need the outer ()s

ojhunt reviewed (commented):
mostly just nits

ojhunt commented on libunwind/src/Registers.hpp:
```diff
@@ -1855,7 +1856,16 @@ class _LIBUNWIND_HIDDEN Registers_arm64 {
   static const char *getRegisterName(int num);
-  void        jumpto() { __libunwind_Registers_arm64_jumpto(this); }
+#ifdef _LIBUNWIND_TRACE_RET_INJECT
+  // clang-format off
```
remove the clang format commands, just accept the misery :D

ojhunt commented on libunwind/src/libunwind_ext.h:
```diff
@@ -30,6 +30,14 @@ extern int __unw_get_reg(unw_cursor_t *, unw_regnum_t, unw_word_t *);
 extern int __unw_set_reg(unw_cursor_t *, unw_regnum_t, unw_word_t);
 extern int __unw_set_fpreg(unw_cursor_t *, unw_regnum_t, unw_fpreg_t);
+#ifdef _LIBUNWIND_TRACE_RET_INJECT
+__attribute__((noinline, disable_tail_calls))
```
rather than duplicating this ifdef pair, define an attribute macro once, something `_LIBUNWIND_TRACE_FN` -- I'm terrible at naming, please don't just use this name verbatim. Possibly ping @ldionne to see if he has an opinion

ojhunt reviewed (changes requested):
There needs to be a better approach to determining how many frames to skip. My intuition is that the easiest way would be to walk up the stack until you find one of the initial entry points. That's annoying, but doable.

ojhunt commented:
Dynamic search for this would be something like

```cpp
bool inbounds_of_unwind_raise_exception(const void*);
template<unsigned N> int _frames_to_unwind_raise_exception() {
  const void *current = __builtin_return_address(N);
  if (inbounds_of_unwind_raise_exception(current)) // this would actually be a bounds check
    return N;
  if constexpr (N > 8) {
    // we should really have reached it by now, and 8 is a power
    // of two so will be presumed to have a hardware related reason :D :D :D
    // Always use 2^^X +/-{0,1} so they seem intentional #BadTeacher
    return -1; // Any "we can't work it out" will do
  } else {
    return _frames_to_unwind_raise_exception<N + 1>();
  }
}

int frames_to_unwind_raise_exception() {
    return _frames_to_unwind_raise_exception<0>();
}
```

you need to add new symbols to make the bounds check possible.

ldionne reviewed (commented):
Can you add a test like we discussed? It might have to be in LLDB but that's already better than nothing.

ojhunt commented on libunwind/src/libunwind_ext.h:
```diff
@@ -30,6 +30,10 @@ extern int __unw_get_reg(unw_cursor_t *, unw_regnum_t, unw_word_t *);
 extern int __unw_set_fpreg(unw_cursor_t *, unw_regnum_t, unw_fpreg_t);
+_LIBUNWIND_TRACE_NO_INLINE
+extern int __unw_resume_with_frames_walked(unw_cursor_t *, unsigned);
+// Legacy function. Do not use.
```
What is the legacy function?

ojhunt commented on libunwind/src/Registers.hpp:
```diff
@@ -1855,7 +1856,14 @@ class _LIBUNWIND_HIDDEN Registers_arm64 {
+    __libunwind_Registers_arm64_jumpto(this, walkedFrames);
+  }
+#else
+  void jumpto() { __libunwind_Registers_arm64_jumpto(this, 0); }
```
indentation

ojhunt commented on libunwind/src/UnwindCursor.hpp:
```diff
@@ -472,7 +472,10 @@ class _LIBUNWIND_HIDDEN AbstractUnwindCursor {
   }
-  virtual void jumpto() { _LIBUNWIND_ABORT("jumpto not implemented"); }
+  _LIBUNWIND_TRACE_NO_INLINE
+  virtual void jumpto() {
```
I think the line length here makes it reasonable to have the no_inline and function decl on the same line

ojhunt commented on libunwind/src/UnwindCursor.hpp:
```diff
@@ -965,6 +974,7 @@ class UnwindCursor : public AbstractUnwindCursor{
   virtual void        setFloatReg(int, unw_fpreg_t);
   virtual int         step(bool stage2 = false);
   virtual void        getInfo(unw_proc_info_t *);
+  _LIBUNWIND_TRACE_NO_INLINE
```
indent or same line for jumpto, I'll leave it to you to decide which reads better/is clearer :D

ojhunt reviewed (approved):
After a bunch of discussion I've been convinced that the hardcoded frame count is reasonable for now, beyond that there are a bunch of style nits that should be fixed before merging.

medismailben committed:
[libunwind] Fix execution flow imbalance when using C++ Exceptions

This patch should fix tracing support when using C++ Exceptions.

When unwinding C++ Exceptions, libunwind would have a first searching
phase to detect its caller landing pad using the personality function,
and a second phase that returns to the landing pad and restore the
register context by also updating the link register to point to the
landing pad address found inthe first phase.

Since it changes the link register value and returns all the frames
that have been called in libunwind, it causes an imbalance in the execution
flow which breaks Apple Processor Trace analysis and affects its
clients, like Instruments.app.

This patch addresses the issue by generating the right amount of `ret`
instructions for every function called in libcxx & libunwind to
rebalance the execution flow, right before returning to the catch block.

rdar://131181678
rdar://131622012

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>

medismailben commented:
Taking a look now @nico

medismailben commented:
libunwind isn't formatted properly so in my understanding, I don't think we enforce clang-format there.
